src/pudl/helpers.py

```diff
@@ -339,6 +339,229 @@ def is_doi(doi):
+    to ``merge_type`` on the columns specified by the ``shared_merge_cols``
+    and ``temporal_merge_cols`` argument, which list the new temporal columns
+    to merge on as well any additional shared columns. Finally, the datetime
+    column is reconstructed in the output dataframe and named according to the
+    ``new_date_col`` parameter.
+
+    Args:
+        left: The left dataframe in the merge. Typically monthly in our use
+            cases if doing a left merge E.g. ``generation_eia923``.
+            Must contain columns specified by ``left_date_col`` and
+            ``shared_merge_cols`` argument.
+        right: The right dataframe in the merge. Typically annual in our uses
+            cases if doing a left merge E.g. ``generators_eia860``.
+            Must contain columns specified by ``right_date_col`` and and
+            ``shared_merge_cols`` argument.
+        left_date_col: Column in ``left`` containing datetime like data. Default is
+            ``report_date``. Must be convertible to a Datetime using
+            :func:`pandas.to_datetime`
+        new_date_col: Name of the reconstructed datetime column in the output dataframe.
+        right_date_col: Column in ``right`` containing datetime like data. Default is
+            ``report_date``. Must be convertible to a Datetime using
+            :func:`pandas.to_datetime`.
+        shared_merge_cols: The columns to merge on that are shared between both
+            dataframes. Typically ID columns like ``plant_id_eia``, ``generator_id``
+            or ``boiler_id``.
+        temporal_merge_cols: The temporal columns to merge on. Values in this list
+            of columns must be [``year``, ``quarter``, ``month``, ``day``].
+            E.g. if a monthly reported dataframe is being merged onto a daily reported
+            dataframe, then the merge would be performed on ``["year", "month"]``.
+        merge_type: How the dataframes should be merged.
+            Values are ["left", "right", "outer", "inner", "cross"].
+            See :func:`pandas.DataFrame.merge`.
+        report_at_start: Whether the data in the dataframe whose report date is not being
+            kept in the merged output (in most cases the less frequently reported dataframe)
+            is reported at the start or end of the time period e.g. January 1st
+            for annual data.
+        kwargs : Additional arguments to pass to :func:`pandas.DataFrame.merge`.
+
+
+    Returns:
+        Merged contents of left and right input dataframes.
+
+    Raises:
+        ValueError: if ``left_date_col`` or ``right_date_col`` columns are missing from their
+            respective input dataframes.
+        ValueError: if any of the labels referenced in ``shared_merge_cols`` are missing from either
+            the left or right dataframes.
+
+    """
+
+    def separate_date_cols(df, date_col_name):
+        df[date_col_name] = pd.to_datetime(df[date_col_name])
```

Comment:
I decided to convert the column to a string if it's not not already a `datetime` column as well as warn if the type is integer. Checking if it's a string or an integer doesn't work that well because if it's either all strings or it's mixed type (some strings and some integers) then it will be read as an `object`, so it's safest to just convert to a string.